### DISCUSSION TITLE
What is the best practice of using `Json<T>` to parse req.body

### DISCUSSION THREAD

**Discussion by duskmoon314** [rwf2/Rocket#1921] [Questions] [answered]
I'm new to rocket, and trying to build a pastebin example based on a database.

I want to use `{"language": <language-name>, "content": <code-content>}` as req.body, and write code as follow:

```rust
#[derive(Deserialize)]
pub struct ReqCode<'r> {
    pub language: Option<&'r str>,
    pub content: &'r str,
}

#[post("/", data = "<code>", format = "json")]
pub async fn create(
    code: Json<ReqCode<'_>>,
    db: &State<Arc<Rbatis>>,
) -> Result<Json<Uuid>, status::Forbidden<Json<RbatisError>>> {
    let id = Uuid::new_v5(&Uuid::NAMESPACE_OID, code.content.as_bytes());
    match db
        .save(
            &Code {
                id,
                language: code.language.map(String::from),
                content: String::from(code.content),
                created_at: NaiveDateTime::now(),
            },
            &[],
        )
        .await
    {
        Ok(_) => Ok(Json(id)),
        Err(e) => Err(status::Forbidden(Some(Json(e)))),
    }
}
```

And I came across this error when trying a basic test:

```http
POST 127.0.0.1:8000 http/1.1
Content-Type: application/json

{
  "language": "C",
  "content": "printf(\"Hello World\n\");"
}
```

```
POST /pastebin application/json:
   >> Matched: (create) POST /pastebin/ application/json
   >> `Json < ReqCode < '_ > >` data guard failed: Parse("{\n    \"language\": \"C\",\n    \"content\": \"printf(\\\"Hello World\\n\\\");\"\n}", Error("invalid type: string \"printf(\\\"Hello World\\n\\\");\", expected a borrowed string", line: 3, column: 43)).
   >> Outcome: Failure
   >> No 422 catcher registered. Using Rocket default.
   >> Response succeeded.
```

I tried to find the `JSON example on GitHub`, but it has been delated in `v0.5-rc`. And in `v0.4`, it use `String` instead of `&str` which is used on the site:

```rust
use rocket::serde::{Deserialize, json::Json};

#[derive(Deserialize)]
struct Task<'r> {
    description: &'r str,
    complete: bool
}

#[post("/todo", data = "<task>")]
fn new(task: Json<Task<'_>>) { /* .. */ }
```

I tried this and it works:

```rust
#[derive(Deserialize)]
pub struct ReqCode<'r> {
    pub language: Option<&'r str>,
    pub content: String,
}
```

I wonder what is the best practice of using `Json<T>`?

**Comment by jebrosen** [comment, marked as answer]
This is a current limitation of serde's zero-copy deserialization: if the input contains escape sequences, there is nowhere for `serde` to put the decoded string such that an `&'r str` could refer to it. A `Cow<'r, str>` or `String` should work; see also serde-rs/json#742.

> I tried to find the JSON example on GitHub, but it has been delated in v0.5-rc.

It hasn't been deleted entirely, but moved to https://github.com/SergioBenitez/Rocket/blob/v0.5-rc/examples/serialization/src/json.rs . (I think there are still a few broken links floating around that are fixed in the repository but not yet published). That example uses `Cow<'r, str>`, which ought to work for you as well.

**Reply by duskmoon314** [reply]
Thanks for answer, very useful
